File: public/books/earnings-management/code/ch07_mcnichols.py
# =====================================================
# code/ch07_mcnichols.py
# 第 7 章 McNichols (2002) 改进 DD 的 Python 实现
# =====================================================

import pandas as pd
import numpy as np
import statsmodels.api as sm

np.random.seed(2026)

p = pd.read_csv("data/em_panel.csv", low_memory=False)
p = p.sort_values(["gvkey", "fyear"]).reset_index(drop=True)
p["CFO_lag"] = p.groupby("gvkey")["CFO_s"].shift(1)
p["CFO_lead"] = p.groupby("gvkey")["CFO_s"].shift(-1)
p2 = p.dropna(subset=["WC_accr", "CFO_lag", "CFO_s",
                      "CFO_lead", "dSale_s",
                      "PPE_s"]).reset_index(drop=True)

print(f"McNichols 回归可用 firm-year: {len(p2)}")

X = sm.add_constant(p2[["CFO_lag", "CFO_s", "CFO_lead",
                        "dSale_s", "PPE_s"]])
fit = sm.OLS(p2["WC_accr"], X).fit()

print("\n======== 7.1 McNichols pooled 回归系数 ========")
print(fit.summary().tables[1])
print(f"R^2 = {fit.rsquared:.4f}")

p2["DA_mcn"] = fit.resid
print("\n======== 7.2 DA_mcn 描述统计 ========")
print(p2["DA_mcn"].describe().round(4))

p2["rank_mcn"] = (
    p2.groupby("fyear")["DA_mcn"]
      .transform(lambda s: s.abs().rank(pct=True))
)
case = p2.loc[p2["company"].notna(),
              ["company", "fyear", "misstate", "DA_mcn", "rank_mcn"]]
print("\n======== 7.3 案例公司 DA_mcn ========")
print(case.sort_values(["company", "fyear"]).to_string(index=False))

File: public/books/earnings-management/code/ch07_mcnichols.R
# =====================================================
# code/ch07_mcnichols.R
# McNichols (2002) 改进的 DD:
#   WC_accr = b0 + b1*CFO_{t-1} + b2*CFO_t + b3*CFO_{t+1}
#                + b4*dSale_s + b5*PPE_s + e
# DA = |e|，按 firm-year
# =====================================================

suppressPackageStartupMessages({
  library(tidyverse)
  library(broom)
  library(here)
})

set.seed(2026)
source(here::here("code", "_load_panel.R"))

p <- load_em_panel()

p2 <- p |>
  arrange(gvkey, fyear) |>
  group_by(gvkey) |>
  mutate(
    CFO_lag  = lag(CFO_s),
    CFO_lead = lead(CFO_s)
  ) |>
  ungroup() |>
  filter(!is.na(WC_accr), !is.na(CFO_lag), !is.na(CFO_s),
         !is.na(CFO_lead), !is.na(dSale_s), !is.na(PPE_s))

cat(sprintf("McNichols 回归可用 firm-year: %d\n", nrow(p2)))

fit <- lm(WC_accr ~ CFO_lag + CFO_s + CFO_lead + dSale_s + PPE_s,
          data = p2)
cat("\n======== 7.1 McNichols pooled 回归系数 ========\n")
print(tidy(fit), digits = 4)
cat(sprintf("R^2 = %.4f\n", summary(fit)$r.squared))

p2$DA_mcn <- residuals(fit)

cat("\n======== 7.2 DA_mcn 描述统计 ========\n")
print(
  p2 |> summarise(
    n        = n(),
    mean     = mean(DA_mcn),
    median   = median(DA_mcn),
    sd       = sd(DA_mcn),
    abs_mean = mean(abs(DA_mcn))
  ),
  digits = 4
)

case_rank <- p2 |>
  group_by(fyear) |>
  mutate(rank_mcn = percent_rank(abs(DA_mcn))) |>
  ungroup() |>
  filter(!is.na(company)) |>
  select(company, fyear, misstate, DA_mcn, rank_mcn) |>
  arrange(company, fyear)

cat("\n======== 7.3 案例公司 DA_mcn ========\n")
print(case_rank, n = Inf, width = Inf)

write_csv(case_rank, here::here("data", "ch07_case_ranks.csv"))
cat("\n案例公司打分已写出：data/ch07_case_ranks.csv\n")
